src/useFetch.ts

```diff
@@ -1,21 +1,24 @@
 import { useEffect, useState, useCallback, useRef, useContext, useMemo } from 'react'
 import FetchContext from './FetchContext'
-import { HTTPMethod, Options, UseFetch, FetchCommands, DestructuringCommands, UseFetchResult, useFetchArg1 } from "./types"
+import { HTTPMethod, Options, UseFetch, FetchCommands, DestructuringCommands, UseFetchResult, useFetchArg1, UseFetchOptions } from "./types"
 import { invariant, isObject } from './utils'
 
-
-export function useFetch<TData = any>(arg1?: useFetchArg1, arg2?: Options | RequestInit): UseFetch<TData> {
+export function useFetch<TData = any>(route?: string, options?: Options): UseFetch<TData>
+export function useFetch<TData = any>(route?: string, requestInit?: RequestInit): UseFetch<TData>
+export function useFetch<TData = any>(useFetchOptions?: UseFetchOptions, options?: Options): UseFetch<TData>
+export function useFetch<TData = any>(useFetchOptions?: UseFetchOptions, requestInit?: RequestInit): UseFetch<TData>
+export function useFetch<TData = any>(routeOrOptions?: useFetchArg1, optionsOrRequestInit?: Options | RequestInit): UseFetch<TData> {
```

Comment:
I've been thinking about this a lot and I think this is how we want to do it.
```ts

type OptionsNoURLs = {
  onMount?: boolean,
  timeout?: number
} & RequestInit

// No Provider
type URLRequiredOptions = {
  url: string,
  onMount?: boolean,
  timeout?: number,
} & RequestInit

type BaseURLRequiredOptions = {
  baseUrl: string
  onMount?: boolean,
  timeout?: number,
} & RequestInit

type OptionsAsFirstParam = URLRequiredOptions | BaseURLRequiredOptions

// With Provider
type MaybeURLOptions = {
  url?: string,
  onMount?: boolean,
  timeout?: number,
} & RequestInit

type MaybeBaseURLOptions = {
  baseUrl?: string
  onMount?: boolean,
  timeout?: number,
} & RequestInit

type OptionsAsFirstParamWithContext = MaybeURLOptions | MaybeBaseURLOptions

// No <Provider url='example.com' />
export function useFetch<TData = any>(url: string, options?: OptionsNoURLs): UseFetch<TData>
export function useFetch<TData = any>(options: OptionsAsFirstParam): UseFetch<TData>
// With <Provider url='example.com' />
// local url will overwrite the global url
// options should be extended. In future maybe have options callback to completely overwrite options
// i.e. useFetch('ex.com', oldOptions => ({ ...newOptions })) to overwrite
export function useFetch<TData = any>(url?: string, options?: OptionsNoURLs): UseFetch<TData>
export function useFetch<TData = any>(options?: OptionsAsFirstParamWithContext): UseFetch<TData>
```

Reply:
I'm still figuring this out though so bare with me :)

Reply:
@alex-cory you will definitely know better than me.  Take your time, I'm not here to give you stuff to do.  You've done a great job so far and I'd like to help out if I can.

Absolutely no rush from me.

Reply:
Okay, I'm gonna finish the cleanup of the arguments here, but the tests we can keep.  You've inspired me to really clean this code up and make it more maintainable with this. Thank you! Hopefully [this PR](https://github.com/alex-cory/use-http/compare/cleaning-up-useFetch-args) can show how I'm doing it and that it's done in a clean/understandable way. So, if it's okay with you, let's rip this part out of this PR and land the tests. 😊

Reply:
totally fine by me and probably better done by you as you know the full picture more.

Looking forward to seeing the outcome.

Reply:
@alex-cory the PR is now just the added tests